### PR TITLE
fixed update inactivity timeout attribute

### PR DESCRIPTION
**Pull request by YevhenBondarenko**
## Pull Request description

Screenshot from 2023-08-02 22-56-21

When device activity becomes false, we update InactivityTimeout to a greater value, then device activity becomes true, and after the timeout remains true.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

**Comment by YevhenBondarenko** [comment]
@smatvienko-tb plz check!

File: application/src/main/java/org/thingsboard/server/service/state/DefaultDeviceStateService.java [hunk 1]
```diff
@@ -285,7 +284,6 @@ public void onDeviceInactivityTimeoutUpdate(TenantId tenantId, DeviceId deviceId
         log.trace("on Device Activity Timeout Update device id {} inactivityTimeout {}", deviceId, inactivityTimeout);
         DeviceStateData stateData = getOrFetchDeviceStateData(deviceId);
         stateData.getState().setInactivityTimeout(inactivityTimeout);
-        checkAndUpdateState(deviceId, stateData);
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/service/state/DefaultDeviceStateService.java:288]: I am still trying to figure out what this fix is about.

If I increase the inactivity timeout - the state has to be recalculated (same as restarting the application with a new timeout)
If I decrease the inactivity timeout - the state also has to be recalculated (same as application restart)

The only change I can imagine is not to fire `onDeviceActivity` when attribute `inactivityTimeout` was changed by TransportToDeviceActorMsg

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 2]
```diff
@@ -125,4 +140,56 @@ private void process(Map<EntityKeyType, Map<String, TsValue>> latest, long defau
...
+
+        service.deviceStates.put(deviceId, deviceStateData);
+
+        service.onDeviceActivity(tenantId, deviceId, System.currentTimeMillis());
+
+        Mockito.verify(telemetrySubscriptionService, Mockito.times(1)).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.eq(true), Mockito.any());
+
+        Mockito.reset(telemetrySubscriptionService);
+
+        service.onDeviceInactivityTimeoutUpdate(tenantId, deviceId, 1);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:187]: wait 1 ms before

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
It seems it is cut off some functionality. Or not enough information provided what the aim of this PR

**Comment by smatvienko-tb** [comment]
Test cases:
1. short timeout = 10sec. send data. active=true. verify: inactive in 10+ sec (checked by scheduler by DeviceService). Test case: increase timeout to 120sec -> state active immediately. after 120+ sec inactivity since lastActivityTime -> active=false - onInactivity event. Summary: inactivity event will appear twice in this test

2. long timeout = 60 sec. send data. active=true. change inactivityTimeot to 1s. Verify active is false immediately. inactivitiAlarmTime has changed as well

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 3]
```diff
@@ -125,4 +140,81 @@ private void process(Map<EntityKeyType, Map<String, TsValue>> latest, long defau
...
+
+        Thread.sleep(2000);
+
+        service.checkStates();
+
+        Mockito.verify(telemetrySubscriptionService, Mockito.times(1)).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.eq(false), Mockito.any());
+
+        Mockito.reset(telemetrySubscriptionService);
+
+        service.onDeviceInactivityTimeoutUpdate(tenantId, deviceId, 2000);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:207]: we already waited 2000ms before (Thread.sleep). Setting inactivite timeout 2000 have to report Active=false, but test if answering true

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 4]
```diff
@@ -125,4 +140,81 @@ private void process(Map<EntityKeyType, Map<String, TsValue>> latest, long defau
...
+
+        Mockito.verify(telemetrySubscriptionService, Mockito.never()).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.eq(true), Mockito.any());
+
+        service.onDeviceActivity(tenantId, deviceId, System.currentTimeMillis());
+
+        Mockito.verify(telemetrySubscriptionService, Mockito.times(1)).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.eq(true), Mockito.any());
+
+        service.onDeviceInactivityTimeoutUpdate(tenantId, deviceId, 1);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:215]: wait 1ms before

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 5]
```diff
@@ -125,4 +140,81 @@ private void process(Map<EntityKeyType, Map<String, TsValue>> latest, long defau
...
+        var tenantId = new TenantId(UUID.randomUUID());
+        var tpi = TopicPartitionInfo.builder().myPartition(true).build();
+        Mockito.when(partitionService.resolve(ServiceType.TB_CORE, tenantId, deviceId)).thenReturn(tpi);
+
+        var deviceIdInfo = new DeviceIdInfo(tenantId.getId(), null, deviceId.getId());
+
+        Mockito.when(deviceService.findDeviceIdInfos(Mockito.any()))
+                .thenReturn(new PageData<>(List.of(deviceIdInfo), 0, 1, false));
+
+        Method method = AbstractPartitionBasedService.class.getDeclaredMethod("initStateFromDB", Set.class);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:164]: may be just call public `service.onTbApplicationEvent(new PartitionChangeEvent(...));`

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 6]
```diff
@@ -125,4 +140,81 @@ private void process(Map<EntityKeyType, Map<String, TsValue>> latest, long defau
         Assert.assertEquals(5000L, deviceStateData.getState().getInactivityTimeout());
     }
 
+    @Test
+    public void givenIncreaseInactivityTimeoutAndThenStateIsActive() throws Exception {
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:144]: The best approach is to have many small corner cases instead of all-in-one cases with confused names.

General cases: Last Inactivity = 0. LastInactivity > 0

IncreaseInactivity for active device . state == no change
IncreaseInactivity for inactive device. state -> active
IncreaseInactivity a little bit for inactive device. state == no change

Decrease inactivity cases both for Last Inactivity = 0. and LastInactivity > 0

On the timeline, we have key points that can happen in different combinations. The initial state can be both active or inactive for any reason before the test event happen.

`LastInactivityTime ... LastActivityTime ... now ... (LastActivityTime + InactivityTimeout)`

Before your changes, some tests have to fail. After your changes, everything will turn **green**!

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
The solution looks adequate, but the test needs to be simplified to read and understand.
As it is a core functionality, I suggest going through all the states as possible.
For 4 time points, the estimated test count is about `4!=24`. the initial state add x2 more. And LastInactivity can be 0 or less which can add even more cases. 
The test part is the hardest thing in that PR. But the **outcome** will be 100% **solid**.

**Review by ViacheslavKlimov — APPROVED** [review]

**Review by smatvienko-tb — APPROVED** [review]
After all discussions, fixes and improvements the overall changes are clear and tidy:

setLastInactivityAlarmTime behind lastActivityTime. 
The only possible value for setLastInactivityAlarmTime is 0L.

```java
               if (deviceState.getLastInactivityAlarmTime() != 0L && deviceState.getLastInactivityAlarmTime() >= deviceState.getLastActivityTime()) {
                    deviceState.setLastInactivityAlarmTime(0L);
                    save(deviceId, INACTIVITY_ALARM_TIME, 0L);
                }
```
Some tests were added and 2 of them were fixed by this PR
I see no problem with PR

**Review by volodymyr-babak — APPROVED** [review]
LGTM
